mautic/mautic
# Autofill form field with lead field

Hi all,

This PR adds a feature to autofill form field with lead field. Previously, the same result could be achieved by using query string parameters but I felt that this approach is a bit "techie" and most marketing users who should be Mautic's primary audience would struggle and a more "mainstream" implementation was needed. Furthermore, the query string parameter route meant that lead details were exposed on the URL.

This PR added the following option to the form field definition screen. When linked with a lead field and Autofill enabled, the form field will get populated with the lead field value on the landing page.

![autofill](https://cloud.githubusercontent.com/assets/13978633/12397436/13572bec-be05-11e5-9312-0cfedd470617.png)

With respect to the query string parameter feature before this, if a query string parameter value is set for the same field then the query string parameter value would take precedence.

**Please note** that this PR also has the code from https://github.com/mautic/mautic/pull/1257 committed to it then backed out because one of my team member made a mistake with the git branch.

**To test**
1) Apply the code changes
2) Add a is_auto_fill column with type tinyint(1) to the form_fields table. You can do this by running 'php app/console doctrine:migrations:migrate'
3) It may be worth clearing out the cache
4) When sending an email to a lead with landing page link, make sure to view it in a different browser session (incognito mode is useful). This is because if you view the email in the same browser session that is currently logged into Mautic then the lead will not be loaded properly.
5) Use both this autofill facility and the previous query string parameter facility to prepopulate the form. You should see the value used in query string used.

After the change, you should be able to use the functionality described above

While testing on 1.2.4, I noticed that if you try to update an existing field with the autofill option and save then it will not save to db. I tried saving the other on/off options like "save values" and these do not persist either so I assume it's a different issue elsewhere. Creating a new field will work normally.
